### PR TITLE
fix(core): added disable functions for previous/next buttons in the calendar

### PR DESCRIPTION
**Pull request by Lukakatche**
## Related Issue(s)

closes https://github.com/sap/fundamental-ngx/issues/5727

## Description

added 2 disable functions for previous and next buttons

### CODE REVIEW

**Review by InnaAtanasova — APPROVED** [review]

File: libs/core/src/lib/calendar/calendar-header/calendar-header.component.ts [hunk 1]
```diff
@@ -212,6 +218,23 @@ export class CalendarHeaderComponent<D> implements OnDestroy, OnInit, OnChanges
     /** Get information about amount of years displayed at once on year view  */
     private _amountOfYearsPerPeriod = 1;
 
+    /**
+     * Function used to disable previous button in the calendar header.
+     */
+
+    @Input()
+    set previousButtonDisableFunction(disableFunc: (date?: D) => boolean) {
+        this.previousButtonDisabled = disableFunc();
```
Comment (platon-rov) [inline libs/core/src/lib/calendar/calendar-header/calendar-header.component.ts:227]: Current date not passed to the function, but i guess this is the idea of this feature. Also the function should be invoked every time new date is selected, not only once when binding.

**Review by platon-rov — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/calendar/calendar-header/calendar-header.component.ts (hunk 1 above)
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar-header/calendar-header.component.ts:227]: Agree with @platon-rov:
1. Checks should be applied whenever bindings are changed, date is changed or date view is changed
2. As a context it would be better to pass selected date and selected view type 
3. Arguments should be passed to the function
4. You should also check if provided value is actually a function
5. Extract type of the function to a separate type, that can be used outside of the component

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 2]
```diff
@@ -232,6 +232,18 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
     @Input()
     disableFunction: (date: D) => boolean = () => false;
 
+    /**
+     * Function used to disable previous button in the calendar header.
+     */
+    @Input()
+    previousButtonDisableFunction: (date?: D) => boolean = () => false;
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.ts:239]: I don't think default value is needed, you should just handle the case provided value is not a function in `calnedar-header`

File: apps/docs/src/app/core/component-docs/calendar/calendar-docs.component.html [hunk 3]
```diff
@@ -164,3 +164,15 @@
     <fd-calendar-form-example></fd-calendar-form-example>
 </component-example>
 <code-example [exampleFiles]="calendarFormSource"></code-example>
+
+<fd-docs-section-title id="reactiveForm" componentName="calendar">
```
Comment (dmitry-stepanenko) [inline apps/docs/src/app/core/component-docs/calendar/calendar-docs.component.html:168]: wrong id

File: apps/docs/src/app/core/component-docs/calendar/calendar-docs.component.html [hunk 4]
```diff
@@ -164,3 +164,15 @@
...
+    Calendar with disabled previous/next buttons</fd-docs-section-title
+>
+<description
+    >You can disable previous/next buttons on the calendar by passing in previousButtonDisableFunction or
+    nextButtonDisableFunction.</description
+>
+<component-example>
+    <fd-calendar-disabled-navigation-buttons-example></fd-calendar-disabled-navigation-buttons-example>
+</component-example>
+<code-example [exampleFiles]="calendarFormSource"></code-example>
```
Comment (dmitry-stepanenko) [inline apps/docs/src/app/core/component-docs/calendar/calendar-docs.component.html:178]: wrong source files

File: apps/docs/src/app/core/component-docs/calendar/examples/calendar-disabled-navigation-buttons-example/calendar-disabled-navigation-buttons-example.html [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<form class="flex-form">
+    <div>
+        <fd-calendar
+            calType="single"
+            [startingDayOfWeek]="1"
+            [previousButtonDisableFunction]="previousButtonDisableFunction"
+            [nextButtonDisableFunction]="nextButtonDisableFunction"
+        ></fd-calendar>
+    </div>
+</form>
```
Comment (dmitry-stepanenko) [inline apps/docs/src/app/core/component-docs/calendar/examples/calendar-disabled-navigation-buttons-example/calendar-disabled-navigation-buttons-example.html:10]: why do we need such complex layout? can it be just the `<fd-calendar>` in root?

File: e2e/wdio/core/pages/calendar.po.ts [hunk 6]
```diff
@@ -18,6 +18,7 @@ export class CalendarPo extends CoreBaseComponentPo {
     mondayCalendar = 'fd-calendar-monday-start-example ';
     internationalCalendar = 'fd-calendar-i18n-example ';
     reactiveCalendarExamples = 'fd-calendar-form-example ';
+    disabledNavigationButtonsCalendar = 'fd-calendar-disabled-navigation-buttons-example';
```
Comment (dmitry-stepanenko) [inline e2e/wdio/core/pages/calendar.po.ts:21]: it's not used in e2e tests, so why is this added?

File: libs/core/src/lib/calendar/calendar.component.html [hunk 7]
```diff
@@ -8,6 +8,8 @@
     (closeClicked)="closeClicked.emit()"
     (nextClicked)="handleNextArrowClick()"
     (previousClicked)="handlePreviousArrowClick()"
+    [previousButtonDisabled]="previousButtonDisableFunction(selectedDate)"
+    [nextButtonDisabled]="nextButtonDisableFunction(selectedDate)"
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.html:12]: This will be invoked on every CD cycle. Any reasons for this? 
According to my comment above:

1. Checks should be applied whenever bindings are changed, date is changed or date view is changed
2. As a context it would be better to pass selected date and selected view type
3. Arguments should be passed to the function
4. You should also check if provided value is actually a function
5. Extract type of the function to a separate type, that can be used outside of the component

File: apps/docs/src/app/core/component-docs/calendar/examples/calendar-disabled-navigation-buttons-example/calendar-disabled-navigation-buttons-example.component.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+@Component({
+    selector: 'fd-calendar-disabled-navigation-buttons-example',
+    templateUrl: 'calendar-disabled-navigation-buttons-example.html'
+})
+export class CalendarDisabledNavigationButtonsExampleComponent {
+    /**
+     * Function used to disable previous button in the calendar header.
+     */
+    previousButtonDisableFunction(): boolean {
```
Comment (dmitry-stepanenko) [inline apps/docs/src/app/core/component-docs/calendar/examples/calendar-disabled-navigation-buttons-example/calendar-disabled-navigation-buttons-example.component.ts:11]: You should add some logic, so it's clear how to use these functions. Would also be nice to see different use cases for prev and next buttons. And please, use typed arrow functions:
```typescript
import { DisableHeaderButtonFunc } from '@fundamental-ngx/core/calendar';
...
export class ExampleComponent {

previousButtonDisableFunction: DisableHeaderButtonFunc = () => { .... };
nextButtonDisableFunction: DisableHeaderButtonFunc = () => { .... };

```

**Review by dmitry-stepanenko — COMMENTED** [review]
@Lukakatche my comments above are not really addressed. Also noticed few issues with added docs

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Let's add the inputs to datepicker and datetimepicker as well, most devs will not be using the calendar component on its own

File: libs/core/src/lib/calendar/calendar-header/calendar-header.component.ts [hunk 9]
```diff
@@ -24,6 +24,12 @@ import { CalendarCurrent } from '../models/calendar-current';
 import { CalendarYearGrid } from '../models/calendar-year-grid';
 import { CalendarService } from '../calendar.service';
 
+export type NavigationButtonDisableFunction<D> = (
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar-header/calendar-header.component.ts:27]: why this type is declared inside header component, where it's not used? Let's move it to separate file in models folder

File: libs/core/src/lib/calendar/calendar-header/calendar-header.component.html [hunk 10]
```diff
@@ -66,7 +66,8 @@
                 fdType="transparent"
                 glyph="slim-arrow-right"
                 [attr.id]="_nextButtonId"
-                [attr.aria-disabled]="false"
+                [attr.aria-disabled]="nextButtonDisabled ? 'true' : 'false'"
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar-header/calendar-header.component.html:69]: same thing with aria-disabled not needed

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 11]
```diff
@@ -212,12 +212,36 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
...
+    /**
+     * @hidden
+     */
+    get currentlyDisplayed(): CalendarCurrent {
+        return this._currentlyDisplayed;
+    }
+
+    set currentlyDisplayed(currentlyDisplayed: CalendarCurrent) {
+        this._currentlyDisplayed = currentlyDisplayed;
+        if (this.previousButtonDisableFunction && this.nextButtonDisableFunction) {
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.ts:224]: remove the check condition from this setter and apply it inside `_setNavigationButtonsStates` function. Additionally, you should check for presence of `previousButtonDisableFunction` and `nextButtonDisableFunction` functions apart

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 12]
```diff
@@ -232,6 +256,58 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
...
+    set nextButtonDisableFunction(disableFunc: NavigationButtonDisableFunction<D>) {
+        if (typeof disableFunc === 'function') {
+            this._nextButtonDisableFunction = disableFunc;
+            if (this.currentlyDisplayed) {
+                this.nextButtonDisabled = this.nextButtonDisableFunction(
+                    this.selectedDate,
+                    this.currentlyDisplayed,
+                    this.activeView
+                );
+            }
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.ts:305]: just wondering, wouldn't it be more concise to have just check for changed inputs inside `onChanges`?

It can look like this:
```typescript
ngOnChanges(changes: SimpleChanes): void {
  if ('nextButtonDisableFunction' in changes || 'previousButtonDisableFunction' in changes || 'currentlyDisplayed' in changes) {
      this. _setNavigationButtonsStates();
  }
}
```

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 13]
```diff
@@ -643,4 +727,19 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
...
+        this.previousButtonDisabled = this.previousButtonDisableFunction(
+            this.selectedDate,
+            this.currentlyDisplayed,
+            this.activeView
+        );
+        this.nextButtonDisabled = this.nextButtonDisableFunction(
+            this.selectedDate,
+            this.currentlyDisplayed,
+            this.activeView
+        );
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.ts:742]:
```suggestion
        this.previousButtonDisabled = typeof previousButtonDisableFunction === 'function' && this.previousButtonDisableFunction(
            this.selectedDate,
            this.currentlyDisplayed,
            this.activeView
        );
        this.nextButtonDisabled = typeof this.nextButtonDisableFunction === 'function' && this.nextButtonDisableFunction(
            this.selectedDate,
            this.currentlyDisplayed,
            this.activeView
        );
```

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 14]
```diff
@@ -206,18 +208,51 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
...
+     * @hidden
+     */
+    get currentlyDisplayed(): CalendarCurrent {
+        return this._currentlyDisplayed;
+    }
+
+    set currentlyDisplayed(currentlyDisplayed: CalendarCurrent) {
+        this._currentlyDisplayed = currentlyDisplayed;
+    }
+
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.ts:245]: in the recent changes you removed all logic from the setter. at this point we can work with `_currentlyDisplayed` directly as it is in main

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 15]
```diff
@@ -286,6 +321,16 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
         }
     }
 
+    ngOnChanges(changes: SimpleChanges): void {
```
Comment (dmitry-stepanenko) [inline libs/core/src/lib/calendar/calendar.component.ts:324]: add `@hidden`

File: apps/docs/src/app/core/component-docs/calendar/examples/calendar-disabled-navigation-buttons-example/calendar-disabled-navigation-buttons-example.component.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        activeView: FdCalendarView
+    ): boolean {
+        if (date.month === 1) {
+            return true;
+        }
+        if (activeView === 'year') {
+            return true;
+        }
+        return currentlyDisplayedDate?.year > 2023;
+    }
```
Comment (dmitry-stepanenko) [inline apps/docs/src/app/core/component-docs/calendar/examples/calendar-disabled-navigation-buttons-example/calendar-disabled-navigation-buttons-example.component.ts:51]: sorry, I didn't notice this before: don't use hardcoded dates, because this example will stop working in 2024
```suggestion
     * will disable choosing dates before current year also if we are in year view or if we selected January
     */
    previousButtonDisableFunction(
        date: FdDate,
        currentlyDisplayedDate: CalendarCurrent,
        activeView: FdCalendarView
    ): boolean {
        if (date.month === 1) {
            return true;
        }
        if (activeView === 'year') {
            return true;
        }
        return currentlyDisplayedDate?.year < new Date().getFullYear();
    }

    /**
     * Function used to disable next button in the calendar header.
     * @param date selected date
     * @param currentlyDisplayedDate currently displayed date
     * @param activeView current view of calendar
     *  will disable choosing dates after next year, also if we are in year view or if we selected January
     */
    nextButtonDisableFunction(
        date: FdDate,
        currentlyDisplayedDate: CalendarCurrent,
        activeView: FdCalendarView
    ): boolean {
        if (date.month === 1) {
            return true;
        }
        if (activeView === 'year') {
            return true;
        }
        return currentlyDisplayedDate?.year > (new Date().getFullYear() + 1);
    }
```

**Review by dmitry-stepanenko — COMMENTED** [review]
looks much better. still noticing few minor things. and this comment is not addressed https://github.com/SAP/fundamental-ngx/pull/7574#discussion_r800566995

**Review by dmitry-stepanenko — APPROVED** [review]

**Review by mikerodonnell89 — APPROVED** [review]

File: libs/core/src/lib/calendar/calendar.component.ts [hunk 17]
```diff
@@ -218,6 +238,10 @@ export class CalendarComponent<D> implements OnInit, ControlValueAccessor, Valid
     /** @hidden */
     private _adapterStartingDayOfWeek: DaysOfWeek;
 
+    public previousButtonDisabled: boolean;
```
Comment (platon-rov) [inline libs/core/src/lib/calendar/calendar.component.ts:241]: `@hidden`?

**Review by platon-rov — COMMENTED** [review]
Minor items.

**Review by platon-rov — APPROVED** [review]
